l3kernel/l3candidates.dtx

```diff
@@ -2006,6 +2061,117 @@
 %    \end{macrocode}
 % \end{macro}
 %
+%
+% \begin{variable}{\c_sys_shell_escape_int}
+%   Expose the engine's shell escape status to the user.
+%    \begin{macrocode}
+\int_const:Nn \c_sys_shell_escape_int
+  {
+    \sys_if_engine_luatex:TF
+      {
+        \luatex_directlua:D
+          {
+            tex.sprint((status.shell_escape~or~os.execute()) .. " ")
+          }
+      }
+      {
+        \pdftex_shellescape:D
+      }
+  }
+%    \end{macrocode}
+% \end{variable}
+%
+% \begin{macro}[EXP, pTF]{\sys_if_shell:}
+%   Performs a check for whether shell escape is enabled.  This will
+%   return true if either of restricted or unrestircted shell escape
+%   is enabled.
+%    \begin{macrocode}
+\prg_new_conditional:Nnn \sys_if_shell: { p , T , F , TF }
+  {
+    \if_int_compare:w \c_sys_shell_escape_int = 0 ~
+      \prg_return_false:
+    \else:
+      \prg_return_true:
+    \fi:
+  }
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{macro}[EXP, pTF]{\sys_if_shell_unrestricted:}
+%   Performs a check for whether \emph{unrestricted} shell escape is
+%   enabled.
+%    \begin{macrocode}
+\prg_new_conditional:Nnn \sys_if_shell_unrestricted: { p , T , F , TF }
+  {
+    \if_int_compare:w \c_sys_shell_escape_int = 1 ~
+      \prg_return_true:
+    \else:
+      \prg_return_false:
+    \fi:
+  }
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{macro}[EXP, pTF]{\sys_if_shell_unrestricted:}
+%   Performs a check for whether \emph{restricted} shell escape is
+%   enabled.  This will return false if unrestricted shell escape is
+%   enabled.  Unrestricted shell escape is not considered a superset
+%   of restricted shell escape in this case.  To find whether any
+%   shell escape is enabled use \cs{sys_if_shell:}.
+%    \begin{macrocode}
+\prg_new_conditional:Nnn \sys_if_shell_restricted: { p , T , F , TF }
+  {
+    \if_int_compare:w \c_sys_shell_escape_int = 2 ~
+      \prg_return_true:
+    \else:
+      \prg_return_false:
+    \fi:
+  }
+%    \end{macrocode}
+% \end{macro}
+%
+%    \begin{macrocode}
+\tex_chardef:D \c_eighteen = 18 ~
+%    \end{macrocode}
+%
+% \begin{macro}{\sys_shell_now:n}
```

Comment:
In cases with engine-dependent definitions we've so far preferred to do the engine test outside the definition to avoid repeating the test multiple times.